components/script/dom/teeunderlyingsource.rs

```diff
@@ -0,0 +1,212 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::cell::Cell;
+use std::rc::Rc;
+
+use dom_struct::dom_struct;
+use js::jsapi::{Heap, Value};
+use js::jsval::UndefinedValue;
+use js::rust::HandleValue as SafeHandleValue;
+
+use super::bindings::root::MutNullableDom;
+use super::types::{ReadableStream, ReadableStreamDefaultReader};
+use crate::dom::bindings::import::module::Error;
+use crate::dom::bindings::reflector::{reflect_dom_object, DomObject, Reflector};
+use crate::dom::bindings::root::Dom;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::promise::Promise;
+use crate::dom::readablestreamdefaultreader::ReadRequest;
+use crate::dom::teereadrequest::TeeReadRequest;
+use crate::script_runtime::CanGc;
+
+pub enum TeeCancelAlgorithm {
+    Cancel1Algorithm,
+    Cancel2Algorithm,
+}
+
+#[dom_struct]
+pub struct TeeUnderlyingSource {
```

Comment:
We should add a doc for https://streams.spec.whatwg.org/#abstract-opdef-readablestreamdefaulttee, 
and when we implement `tee` for byte streams we're going to have to rename this one to `DefaultTeeUnderlyingSource` to differentiate it from the stuff at https://streams.spec.whatwg.org/#abstract-opdef-readablebytestreamtee